lib/puppet_x/puppetlabs/powershell/powershell_manager.rb

```diff
@@ -131,14 +131,44 @@ def self.init_path
         "\"#{path}\""
       end
 
-      def make_ps_code(powershell_code, timeout_ms = nil, working_dir = nil)
+      def make_ps_code(powershell_code, timeout_ms = nil, working_dir = nil, environment_variables = [])
         begin
           timeout_ms = Integer(timeout_ms)
           # Lower bound protection. The polling resolution is only 50ms
           if (timeout_ms < 50) then timeout_ms = 50 end
         rescue
           timeout_ms = 300 * 1000
         end
+
+        # environment values passed in need to be converted to a PowerShell Hashtable
+        # environment parsing is based on the Puppet exec equivalent code
+        # https://github.com/puppetlabs/puppet/blob/a9f77d71e992fc2580de7705847e31264e0fbebe/lib/puppet/provider/exec.rb#L35-L49
+        environment = {}
+        if envlist = environment_variables
+          envlist = [envlist] unless envlist.is_a? Array
+          envlist.each do |setting|
+            if setting =~ /^(\w+)=((.|\n)+)$/
+              env_name = $1
+              value = $2
+              if environment.include?(env_name) || environment.include?(env_name.to_sym)
+                Puppet.warning("Overriding environment setting '#{env_name}' with '#{value}'")
+              end
+              environment[env_name] = value
+            else
+              Puppet.warning("Cannot understand environment setting #{setting.inspect}")
+            end
+          end
+        end
+        exec_environment_variables = '@{'
+        environment.each do |name,value|
+          # Powershell escapes single quotes inside a single quoted string by just adding 
+          # another single quote i.e. a value of foo'bar turns into 'foo''bar' when single quoted
+          ps_name = name.gsub('\'','\'\'')
+          ps_value = value.gsub('\'','\'\'')
+          exec_environment_variables += " '#{ps_name}' = '#{ps_value}';"
```

Comment:
Just a thought, is this executed in Puppet or PowerShell? Shouldn't this use `$env:` on the front?